gen/map.go

```diff
@@ -190,6 +191,83 @@ func (m *mapGenerator) Reader(g Generator, spec *compile.MapSpec) (string, error
 	return name, wrapGenerateError(spec.ThriftName(), err)
 }
 
+// Equals generates a function to compare maps of the given type
+//
+// func $name(lhs, rhs $mapType) bool {
+//      ...
+// }
+//
+// And returns its name.
+func (m *mapGenerator) Equals(g Generator, spec *compile.MapSpec) (string, error) {
+	name := "_" + valueName(spec) + "_Equals"
+	if m.HasEquals(name) {
+		return name, nil
+	}
+
+	err := g.DeclareFromTemplate(
+		`
+            <$mapType := typeReference .Spec>
+
+            <$lhs := newVar "lhs">
+            <$rhs := newVar "rhs">
+            <$i := newVar "i">
+            <$j := newVar "j">
+            <$lk := newVar "lk">
+            <$lv := newVar "lv">
+            <$rk := newVar "rk">
+            <$rv := newVar "rv">
+            <$ok := newVar "ok">
+			func <.Name>(<$lhs>, <$rhs> <$mapType>) bool {
+				if len(<$lhs>) != len(<$rhs>) {
+					return false
+				}
+
+				<if isHashable .Spec.KeySpec>
+					for <$lk>, <$lv> := range <$lhs> {
+						<$rv>, <$ok> := <$rhs>[<$lk>]
+						if !<$ok> {
+							return false
+						}
+						if !(<equals .Spec.ValueSpec $lv $rv>) {
+							return false
+						}
+					}
+				<else>
+					// Note if keys are not hashable then this is O(n^2) in time complexity.
+					for _, <$i> := range <$lhs> {
+						<$lk> := <$i>.Key
+						<$lv> := <$i>.Value
+						<$ok> := false
+						for _, <$j> := range <$rhs> {
+							<$rk> := <$j>.Key
+							<$rv> := <$j>.Value
+							if !(<equals .Spec.KeySpec $lk $rk>) {
+								continue
+							}
+							if !(<equals .Spec.ValueSpec $lv $rv>) {
+								continue
```

Comment:
can't we just return false if the key matches but the value doesn't? caveat being that behavior is undefined if the map contract (only one instance of each key) isn't satisfied.